Add targeted attack methods.

guangzhuwu opened the pull request:
1. Add an new class "Adversary", which defines a class that contains the original object, the target and the adversarial example.
2. Modify _apply method of IteratorGradientSignAttack and GradientSignAttack to support targeted attack.

guangzhuwu committed:
Add targeted attack methods.

guangzhuwu committed:
Add targeted attack methods

gx-wind commented on fluid/adversarial/mnist_tutorial_fgsm.py:
```diff
@@ -76,10 +76,11 @@ def main():
-        plt.imshow(n[0][0], cmap='Greys_r')
-        plt.show()
-        #np.save('adv_img', adv_img)
+        adversary = att(Adversary(data))
```
The type of data is [(np.numpy, int)]，data[0][0] represents the image and data[0][1] represent the label, so it should be like this?
```python
adversary = att(Adversary(data[0][0], data[0][1]))
```

guangzhuwu replied:
Yep!

guangzhuwu committed:
Bug fix:
1. target_label.setter
2. Adversary(data) should be Adversary(data[0][0], data[0][1])

guangzhuwu committed:
Fix some PEP8 warning.

gx-wind commented on fluid/adversarial/advbox/attacks/gradientsign.py:
```diff
@@ -2,37 +2,58 @@
         for epsilon in epsilons:
-            adv_img = image_label[0][0].reshape(
-                gradient_sign.shape) + epsilon * gradient_sign
+            adv_img = adversary.original + epsilon * gradient_sign
```
The shape of adversary.original and gradient_sign may be not same.For example, in mnist dataset, the shape of adversary.original is (784,), and the shape of gradient_sign is (1, 1, 28, 28), then it will throw an error:
```sh
ValueError: operands could not be broadcast together with shapes (784,) (1,1,28,28)
```

guangzhuwu committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/models into develop

guangzhuwu commented:
The shape bug is fixed.

guangzhuwu committed:
Fix shape bug.

guangzhuwu committed:
Fix gradientsing attack bug.

guangzhuwu committed:
delete assert adversarial_example.shape == self.__original.shape

gx-wind reviewed (approved):
LGTM!
